daemon/peer.c

```diff
@@ -911,11 +911,18 @@ static void their_htlc_added(struct peer *peer, struct htlc *htlc,
 
 	//FIXME: dirty trick to retrieve unexported state
 	memcpy(&pk, peer->dstate->secret, sizeof(pk));
+
 	packet = parse_onionpacket(peer,
 				   htlc->routing, tal_count(htlc->routing));
-	if (packet)
-		step = process_onionpacket(packet, packet, &pk, htlc->rhash.u.u8,
-					   sizeof(htlc->rhash));
+	if (packet) {
+		u8 shared_secret[32];
+
+		if (onion_shared_secret(shared_secret, packet, &pk))
```

Comment:
If I see this correctly this is not communicating with the HSM, and due to the HSM communication being async we'll need to touch this later anyway, to be callback based when we get the ECDH result back. So I don't exactly see how this is supposed to work.